dkc/core/rest/file.py

```diff
@@ -31,12 +33,18 @@ class Meta:
 class FileViewSet(ModelViewSet):
     queryset = File.objects.all()
 
-    permission_classes = [HasAccess]
+    permission_classes = [HasAccess & IsAuthenticatedOrReadOnly]
```

Comment:
I believe `IsAuthenticatedOrReadOnly` logic is redundant with the `perform_create` checks as it is not possible to give the "anonymous user" write access to anything.  I think if we wanted to make this explicit we should change `HasAccess` to
```python
HasAccess = (IsReadOnlyEndpoint & IsReadable) | (IsWriteable & IsAuthenticated)
```

Reply:
I'm happy to make that change. The reason I added this here is because it causes the file creation endpoint to return 401 for anonymous users, which is desired, whereas without it it returns 403 because it falls back to object-level permission checks.

Reply:
That makes sense.  I think I would prefer it on the default permissions class rather than added everywhere it's used.

Reply:
I changed it to 
```python
HasAccess = (IsReadOnlyEndpoint & IsReadable) | (IsAuthenticated & IsWriteable)
```

But that gives 403s instead of 401s. I will try and figure out why that's happening.

Reply:
The explanation is [here](https://www.django-rest-framework.org/api-guide/permissions/#how-permissions-are-determined)

Reply:
Nevermind, that was a red herring. The real issue was that `IsReadOnlyEndpoint` was using `has_object_permission` but it really needs to use `has_permission`, which also gets called on create endpoints.

Reply:
Ah.  That makes sense.